src/main/java/org/folio/rest/impl/VendorHelper.java

```diff
@@ -0,0 +1,194 @@
+          .collect(toList());
+        ids.stream()
+          .filter(id -> !vendorsIds.contains(id))
+          .forEach(id -> errors.add(POL_ACCESS_PROVIDER_NOT_FOUND.toError().withAdditionalProperty(ID, id)));
+        return buildErrorsObject(errors);
+      }).thenAccept(future::complete)
+        .exceptionally(t -> {
+          logger.error("Failed to validate access provider's status", t);
+          future.completeExceptionally(t);
+          return null;
+        });
+    } else {
+      future.complete(buildErrorsObject(errors));
+    }
+    return future;
+  }
+
+  /**
+   * Builds {@link Errors} object based on list of {@link Error}
+   *
+   * @param errors list of {@link Error}
+   * @return {@link Errors} object with list of {@link Error}
+   */
+  private Errors buildErrorsObject(List<Error> errors) {
+    return new Errors()
+      .withErrors(errors)
+      .withTotalRecords(errors.size());
+  }
+
+  /**
+   * Creates {@link Error} with id of corresponding vendor/access provider
+   *
+   * @param id vendor's/access provider's identifier
+   * @param errorCodes error code
+   * @return {@link Error} with id of failed vendor/access provider
+   */
+  private Error createErrorWithId(ErrorCodes errorCodes, String id) {
+    return errorCodes.toError()
+      .withAdditionalProperty(ID, id);
+  }
+
+  /**
+   * Retrieves vendor as json file
+   *
+   * @param vendorId vendor's id
+   * @return CompletableFuture with {@link JsonObject} representation of vendor
+   */
+  private CompletableFuture<JsonObject> getVendorJsonById(String vendorId) {
+    return handleGetRequest(VENDOR_STORAGE_VENDORS + vendorId, httpClient, ctx, okapiHeaders, logger);
+  }
+
+  /**
+   * Retrieves list of access providers
+   *
+   * @param accessProviderIds - list of access providers id
+   * @return CompletableFuture with {@link List<JsonObject>} representation of vendor
+   */
+  private CompletableFuture<List<JsonObject>> getAccessProvidersByIds(List<String> accessProviderIds) {
+    CompletableFuture<List<JsonObject>> future = new CompletableFuture<>();
+    String query = convertIdsToCqlQuery(accessProviderIds);
+    String endpoint = String.format(VENDORS_WITH_QUERY_ENDPOINT, accessProviderIds.size(), lang, encodeQuery(query, logger));
+    handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+      .thenApply(vendors -> vendors.getJsonArray(VENDORS).stream().map(obj -> (JsonObject) obj).collect(toList())).thenAccept(future::complete);
+    return future;
+  }
+
+  /**
+   * Returns list of access providers ids
+   *
+   * @param compPO composite purchase order
+   * @return list of access providers id
+   */
+  private List<String> verifyAndGetAccessProvidersIdsList(CompositePurchaseOrder compPO) {
+    return compPO.getCompositePoLines().stream()
+      .filter(p -> (p.getEresource() != null && p.getEresource().getAccessProvider() != null))
+      .map(p -> p.getEresource().getAccessProvider()).collect(Collectors.toList());
```

Comment:
Please consider either adding `distinct()` or using `toSet()` collector just to not include duplicates to query